main.cpp

```diff
@@ -272,373 +219,208 @@ struct HeapA
-
-    FloatType& apply( void(*func)(float&) );
-
-private:
-    float* floatPtr;
-    FloatType& powInternal( float value );
-};
-
-struct DoubleType
-{
-    DoubleType(double dbl):doublePtr(new double(dbl)){}
-    ~DoubleType()
+    Numeric& operator -=( Type value )
     {
-        delete doublePtr;
-        doublePtr = nullptr;
+        *numberPtr -= value;
+        return *this;
     }
-    operator double() const {return *doublePtr;}
-
-    DoubleType& operator +=( double value );
-    DoubleType& operator -=( double value );
-    DoubleType& operator *=( double value );
-    DoubleType& operator /=( double value );
-
-    DoubleType& pow( double value );
-    DoubleType& pow( const IntType& it );
-    DoubleType& pow( const FloatType& ft );
-    DoubleType& pow( const DoubleType& dt );
 
-    DoubleType& apply( std::function<DoubleType&(double&)> func );
-    DoubleType& apply( void(*func)(double&) );
-
-private:
-    double* doublePtr;
-    DoubleType& powInternal( double value);
-};
-
-struct IntType
-{
-    IntType(int val):intPtr(new int(val)){}
-    ~IntType()
+    Numeric& operator *=( Type value )
     {
-        delete intPtr;
-        intPtr = nullptr;
+        *numberPtr *= value;
+        return *this;
     }
 
-    operator int() const{return *intPtr;}
-
-    IntType& operator +=( int value );
-    IntType& operator -=( int value );
-    IntType& operator *=( int value );
-    IntType& operator /=( int value );
-
-    IntType& pow( int value );
-    IntType& pow( const IntType& it );
-    IntType& pow( const FloatType& ft );
-    IntType& pow( const DoubleType& dt );
-
-    IntType& apply( std::function<IntType&(int&)> func );
-    IntType& apply( void(*func)(int&) );
-
-private:
-    int* intPtr;
-    IntType& powInternal( int value );
-};
-
-struct Point
-{
-    Point(const IntType& x_, const IntType& y_) : x(static_cast<float>(x_)), y(static_cast<float>(y_)){}
-
-    Point(const FloatType& x_, const FloatType& y_) : x(static_cast<float>(x_)), y(static_cast<float>(y_)){}
-
-    Point(const DoubleType& x_, const DoubleType& y_) : x(static_cast<float>(x_)), y(static_cast<float>(y_)){}
-
-  
-    Point& multiply(float m)
+    template<typename ParType>
+    Numeric& operator /=( ParType value )
     {
-        x *= m;
-        y *= m;
+        if constexpr (std::is_same<Type, int>::value)
+        {
+            if constexpr (std::is_same<ParType, int>::value)
+            {
+                if (value == 0)
+                {
+                    std::cout << "error: integer division by zero is an error and will crash the program!" << "\n";
+                    return *this;
+                }
+            }
+            else if (value < std::numeric_limits<ParType>::epsilon())
+            {
+                std::cout << "can't divide integers by zero!\n";
+                return *this;
+            }
+        }
+        else if (value < std::numeric_limits<ParType>::epsilon())
```

Comment:
This should be the template class's type, not the parameter's type.
take a 2nd look at the original instructions:
```
if your template type is an int
    ...
else if it's less than epsilon
```
- [ ] revise